src/main/groovy/org/pih/warehouse/outbound/ImportPackingListItem.groovy

```diff
@@ -0,0 +1,98 @@
+
+    String rowId
+
+    String palletName
+
+    String boxName
+
+    @BindUsing({ obj, source -> Product.findByProductCode(source['product']) })
+    Product product
+
+    String lotNumber
+
+    Date expirationDate
+
+    Location origin
+
+    @BindUsing({ obj, source ->
+        Location internalLocation = Location.findByNameAndParentLocation(source['binLocation'], obj.origin)
+        // If location is not found, but we provided bin location name, it means, the location was not found,
+        // and we want to indicate that, instead of falling back to default (null)
+        // without this, we would then search e.g. for quantity available to promise for a product in the default bin location
+        if (!internalLocation && source['binLocation'] && !source['binLocation'].equalsIgnoreCase(Constants.DEFAULT_BIN_LOCATION_NAME)) {
+            // We want to indicate, that a bin location for given name has not been found.
+            obj.binLocationFound = false
+        }
+        return internalLocation
+     })
+    Location binLocation
+
+    Integer quantityPicked
+
+    @BindUsing({ obj, source -> Person.findPersonByNameOrEmail(source['recipient'])})
+    Person recipient
+
+    /**
+     * Flag to indicate whether binLocation has been found and bound.
+     * We can't rely on binLocation = null, because we accept the null as DEFAULT,
+     * so we want to know if the binLocation is null, because we have not provided (or provided DEFAULT) the bin location name
+     * or the binLocation for such name does not exist.
+     * This can't be done in the BindUsing and rejectValue, because the errors would be cleared before running the validator logic
+     * The flag is set to false if a bin location is not found for given name in the BindUsing of binLocation
+     */
+    boolean binLocationFound = true
+
+
+    static constraints = {
+        rowId(blank: false)
+        palletName(nullable: true)
+        boxName(nullable: true, validator: { String boxName, ImportPackingListItem item ->
+            if (boxName && !item.palletName) {
+                return ['packLevel1.required']
+            }
+            return true
+        })
+        lotNumber(nullable: true, blank: true, validator: { String lotNumber, ImportPackingListItem item ->
+            InventoryItem inventoryItem = item.product?.getInventoryItem(lotNumber)
+            if (inventoryItem) {
+                return true
+            }
+            return ['inventoryItemNotFound', lotNumber, item.product?.productCode]
+        })
+        expirationDate(nullable: true)
+        binLocation(nullable: true)
+        quantityPicked(min: 0, validator: { Integer quantityPicked, ImportPackingListItem item ->
+            if (!item.binLocationFound) {
+                return ['binLocationNotFound']
+            }
+            ProductAvailabilityService productAvailabilityService = Holders.grailsApplication.mainContext.getBean(ProductAvailabilityService)
```

Comment:
I'm not a big fan of having to pull a service in like this to do validation. IMO request objects should be simple and not depend on any other large classes.

My preferred approach to handling situations like this is to add an `ImportPackingListValidator` that can autowire in whatever classes it needs.
- https://docs.spring.io/spring-framework/reference/core/validation/validator.html
- https://stackoverflow.com/questions/41996143/spring-service-method-and-a-complex-validation-logic-rules

That way request objects can act more like simple POJOs. But maybe there's a better way to achieve this via Grails that I don't know about.

Reply:
I'm okay with doing it this way for now, but I would like to brainstorm some potential longterm solutions

Reply:
@EWaterman I didn't want to introduce a validator class or proceed with the validation in the service for this one property, not to mix two approaches - `validator`, and a custom validation in a validator class.
I know injecting a service into a domain is debatable, but since Grails' recommended validation happens in the domain (validator), I've gone this way. 
As I mentioned somewhere above - this doesn't "recreate" the product availability service, or anything like this, because the PA service is singleton, and we always get the same reference no matter which item we are iterating through.